### PR TITLE
[vcpkg] Switch to internal hash algorithms 🐱‍💻

### PR DESCRIPTION
**Pull request by strega-nil**
On non-Windows platforms, there is no standard way to get the hash of an
item -- before this PR, what we did was check for the existence of a few
common utility names (shasum, sha1, sha256, sha512), and then call that
utility on a file we created containing the contents we wish to hash.
This PR adds internal hashers for sha1, sha256, and sha512, and
standardizes the interface to allow anyone to implement hashers in the
future.

These hashers are not extremely optimized, so it's likely that in the
future we could get more optimized, but for now we just call out to
BCryptHasher on Windows, since it's standard and easy to use (and about
2x faster for sha1 and sha256, and 1.5x faster for sha512). However,
they are reasonably fast for being unoptimized. I attempted a few minor
optimizations, which actually made the code slower! So as of right now,
it's implemented as just a basic conversion of the code on Wikipedia to
C++. I have tested these on the standard NIST test vectors (and those
test vectors are located in vcpkg-test/hash.cpp).

### CODE REVIEW

**Comment by ubsan** [comment]
As one can see from [these benchmarks], BCryptHasher takes about 50% of the time as my implementation, which to be fair is the simplest implementation possible. Since BCryptHasher is so much faster, and we're guaranteed to have it on Windows, I chose to keep the BCryptHasher code in on Windows -- on macOS and Linux, I deemed it more important to be able to hash without calling out to the system.

The code as seen right now is incomplete; I will attempt to finish it before the end of today.

note: I attempted some minor optimizations, but they resulted in slower code, so I reverted them.

[these benchmarks]: https://github.com/microsoft/vcpkg/files/3521846/bench.txt

File: toolsrc/include/vcpkg/base/strings.h [hunk 1]
```diff
@@ -12,15 +12,25 @@
...
-    auto to_printf_arg(const T& t) -> decltype(t.to_string())
+    auto to_string(const T& t) -> decltype(t.to_string())
     {
         return t.to_string();
     }
 
+    // first looks up to_string on `T` using ADL; then, if that isn't found,
+    // uses the above definition which returns t.to_string()
+    template<class T, class = std::enable_if_t<!std::is_arithmetic<T>::value>>
+    auto to_printf_arg(const T& t) -> decltype(to_string(t))
```
Comment (ubsan) [inline toolsrc/include/vcpkg/base/strings.h]: Note that this was done for `enum class`es -- to allow them to be printed.

File: toolsrc/include/vcpkg/base/strings.h [hunk 2]
```diff
@@ -12,15 +12,25 @@
...
+    {
+        return to_string(t);
+    }
+
     inline const char* to_printf_arg(const std::string& s) { return s.c_str(); }
 
     inline const char* to_printf_arg(const char* s) { return s; }
 
+    inline const wchar_t* to_printf_arg(const wchar_t* s) { return s; }
```
Comment (ubsan) [inline toolsrc/include/vcpkg/base/strings.h]: This was done because I would like to print `LPCWSTR`s.

**Comment by vicroms** [comment]
LGTM, but I'll admit that I skimmed through the actual algorithms implementation code.

**Pull request merged**